.doom.d/config.org
#+title: Doomemacs Config
* Basic Utility Functions
#+begin_src emacs-lisp :tangle functions.el
;;; $DOOMDIR/functions.el -*- lexical-binding: t; -*-

(defun directory-path-p (path)
  (and (file-exists-p path)
       (file-directory-p path)))

(defun add-project-directories (&rest paths)
  (seq-filter #'directory-path-p paths))
#+end_src

Load the defined functions into main config.el
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(load! "functions")
#+end_src

#+begin_src emacs-lisp :tangle packages.el
;; -*- no-byte-compile: t; -*-
;;; $DOOMDIR/packages.el -*- lexical-binding: t; -*-
#+end_src

* Basic Variables for normal usage

#+begin_src emacs-lisp
(add-to-list 'default-frame-alist '(fullscreen . maximized))

(setq user-full-name                    "Erick G. Islas Osuna"
      user-mail-address                 "erick@tliwaka.com.mx"
      doom-theme                        'doom-molokai
      org-directory                     "~/Documents/Cloud/Notes/"
      display-line-numbers-type         t
      projectile-project-search-path    (add-project-directories "~/Documents/Github"
                                                                 "~/Github"
                                                                 "~/Documents/TradingMonkey"
                                                                 "~/Documents/Proyectos/Flutter"
                                                                 "~/Documents/Proyectos/Clojure"
                                                                 "~/Documents/Proyectos/CommonLisp"
                                                                 "~/Documents/Proyectos/Python"
                                                                 "~/Documents/Proyectos/React"
                                                                 "~/Documents/Proyectos/Go"
                                                                 "~/Documents/Proyectos/VHDL"
                                                                 "~/Documents/Proyectos/Machine Learning"
                                                                 "~/Documents/Proyectos/Serverless"
                                                                 "~/Documents"
                                                                 "~/OpenSource")

      evil-vsplit-window-right          t
      evil-split-window-below           t
      doom-font                         (font-spec :family "Fira Code" :size 12)
      display-line-numbers-type         'relative
      ispell-dictionary                 "es"
      ;; doom-themes-treemacs-theme "all-the-icons"
      )
#+end_src

* Packages

#+begin_src emacs-lisp :tangle packages.el
;; (unpin! treemacs)
;; (unpin! lsp-treemacs)
;; (unpin! lsp-ui)
;; (unpin! lsp-mode)
#+end_src

#+begin_src emacs-lisp :tangle packages.el

(package! treemacs-all-the-icons)
(package! scss-mode)
(package! lsp-metals)
#+end_src

* Po Mode for Docs Translation
#+begin_src emacs-lisp :tangle packages.el
(package! po-mode
  :recipe (:host github :repo "erickisos/po-mode" :files ("*.el" "Makefile")))
#+end_src

#+begin_src emacs-lisp
(use-package! po-mode
  :config
  (map! :map po-mode-map
        :localleader
        :desc "Go to next entry" "n" #'po-next-entry
        :desc "Go to prev entry" "N" #'po-previous-entry
        :desc "Next fuzzy entry" "f" #'po-next-fuzzy-entry
        :desc "Prev fuzzy entry" "F" #'po-previous-fuzzy-entry
        :desc "Next untranslated entry" "u" #'po-next-untranslated-entry
        :desc "Prev untranslated entry" "U" #'po-previous-untranslated-entry
        :desc "Edit msgstr in separated buffer" "e" #'po-edit-msgstr))
#+end_src
* OX Modern CV

#+begin_src emacs-lisp :tangle packages.el
(package! ox-moderncv
  :recipe (:host github :repo "erickisos/org-cv" :files ("*.el")))
#+end_src


* Tree Slide (Slideshow for Geeks?)

#+begin_src emacs-lisp
(use-package! org-tree-slide
  :after org-mode
  :config
  (setq org-tree-slide-slide-in-effect       t
        +org-present-text-scale              2
        org-tree-slide-skip-outline-level    2
        org-tree-slide-fold-subtrees-skipped nil
        org-tree-slide-modeline-display      'outside)
  (map! :map org-tree-slide-mode-map
        :desc "Next slide" "M-n" #'org-tree-slide-move-next-tree
        :desc "Prev slide" "M-p" #'org-tree-slide-move-previous-tree))
#+end_src

* Github Copilot (ECA)

This is the main config to enable ECA

#+begin_src emacs-lisp :tangle packages.el
(package! eca
  :recipe (:host github :repo "editor-code-assistant/eca-emacs" :files ("*.el")))
#+end_src

#+begin_src emacs-lisp
(use-package! eca)
#+end_src

* Github Copilot (Aider  fP)
This is a temporary measure to enable Github Copilot using =copilot.el= while I enable this through ECA.

#+begin_src emacs-lisp :tangle packages.el
(package! copilot
  :recipe (:host github :repo "copilot-emacs/copilot.el" :files ("*.el")))
#+end_src

* Paredit for those who love parentheses

#+begin_src emacs-lisp :tangle packages.el
(package! paredit)
#+end_src

#+begin_src emacs-lisp
(use-package! paredit
  :hook ((clojure-mode . paredit-mode)
         (emacs-lisp-mode . paredit-mode)))
#+end_src

* LSP Configurations

#+begin_src emacs-lisp :tangle packages.el
(unpin! lsp-mode)
(unpin! lsp-ui)
#+end_src

#+begin_src emacs-lisp
(use-package! lsp-mode
  :commands lsp
  :config
  (setq lsp-headerline-breadcrumb-enable t
        lsp-inlay-hints-mode             t
        lsp-lens-enable                  t)
  (advice-add #'lsp-rename
              :after (lambda (&rest _) (projectile-save-project-buffers))))
#+end_src

* Rust Specific configurations

#+begin_src emacs-lisp
(after! lsp-mode
  :config
  (setq lsp-rust-analyzer-display-chaining-hints            t
        lsp-rust-analyzer-display-closure-return-type-hints t
        lsp-rust-analyzer-display-parameter-hints           t))
#+end_src
* Java development

** Gradle Mode!

This is useful for the projects we work with, which are generally tied to Gradle.
#+begin_src emacs-lisp :tangle packages.el
(package! gradle-mode
  :recipe (:host github :repo "erickisos/emacs-gradle-mode" :files ("*.el" "Makefile")))
#+end_src

#+begin_src emacs-lisp
(use-package! gradle-mode
  :after java-mode)
#+end_src

** LSP Java

#+begin_src emacs-lisp
(use-package! lsp-java
  :after java-mode
  :config
  (add-hook 'java-mode-hook (lambda () (gradle-mode 1)))
  (setq lsp-java-references-code-lens-enabled      t
        lsp-java-implementations-code-lens-enabled t
        lsp-java-lens-mode                         t
        lsp-java-format-settings-url               "~/IDEAJavaFormatter.xml"
        lsp-java-vmargs                            '("-XX:+UseParallelGC"
                                                     "-XX:GCTimeRatio=4"
                                                     "-XX:AdaptiveSizePolicyWeight=90"
                                                     "-Dsun.zip.disableMemoryMapping=true"
                                                     "-Xmx8G"
                                                     "-Xms100m"))
  (setq lombok-library-path (concat doom-data-dir "lombok.jar"))
  (unless (file-exists-p lombok-library-path)
    (url-copy-file "https://projectlombok.org/downloads/lombok.jar" lombok-library-path))
  (push (concat "-javaagent:"
                (expand-file-name lombok-library-path))
        lsp-java-vmargs))
#+end_src

** SpringBoot for Java

#+begin_src emacs-lisp
(use-package! lsp-java-boot
  :after lsp-mode
  :preface
  (add-hook 'conf-javaprop-mode-local-vars-hook #'lsp!)
  (add-hook 'yaml-mode-local-vars-hook #'lsp!)
  :config
  (add-hook 'java-mode-local-vars-hook #'lsp-java-boot-lens-mode 'append)
  (add-hook 'yaml-mode-local-vars-hook #'lsp-java-boot-lens-mode 'append))
#+end_src
* Clojure
#+begin_src emacs-lisp
(use-package! clojure-mode
  :config
  (setq clojure-indent-style 'align-arguments))
#+end_src

* Obsidian MD
Before using Obsidian through Emacs, it's also neccessary to mark the iCloud folder as "Mantener Descarga" to avoid issues of refetching files.

#+begin_src emacs-lisp :tangle packages.el
(package! obsidian)
#+end_src

#+begin_src emacs-lisp
(use-package! obsidian
  :config
  (global-obsidian-mode t)
  :custom
  (obsidian-directory "~/Library/Mobile Documents/iCloud~md~obsidian/Documents/ObsNotes")
  (obsidian-daily-notes-directory "Daily"))
#+end_src
* Why This? (GitLens-ish blame mode)

#+begin_src emacs-lisp :tangle packages.el
(package! why-this
  :recipe (:host github :repo "erickisos/emacs-why-this" :files ("*.el" "Makefile")))
#+end_src

#+begin_src emacs-lisp
(use-package! why-this
  :config
  (setq why-this-annotate-enable-heat-map t
        why-this-idle-delay               0.5
        why-this-minimum-column           45)
  (global-why-this-mode))
#+end_src

* Eye Candy

** Pulsing Cursor
#+begin_src emacs-lisp :tangle packages.el
(package! pulsing-cursor
  :recipe (:host github :repo "jasonjckn/pulsing-cursor"))
#+end_src

#+begin_src emacs-lisp
(use-package! pulsing-cursor
  :config
  (pulsing-cursor-mode +1))
#+end_src

* CodeSnap Integration

The codesnap package is now a separate file that gets loaded.

#+begin_src emacs-lisp :tangle packages.el
(package! codesnap
  :recipe (:host github :repo "erickisos/codesnap.el" :files ("*.el")))
#+end_src

Keybindings for codesnap

#+begin_src emacs-lisp
(use-package! codesnap
  :config
  (map! :localleader
        :prefix "CodeSnap" "S"
        :desc "Screenshot selected code" "S x" #'codesnap-selection
        :desc "Show CodeSnap log" "S l" #'codesnap-show-log))
#+end_src


.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(load! "functions")

(add-to-list 'default-frame-alist '(fullscreen . maximized))

(setq user-full-name                    "Erick G. Islas Osuna"
      user-mail-address                 "erick@tliwaka.com.mx"
      doom-theme                        'doom-molokai
      org-directory                     "~/Documents/Cloud/Notes/"
      display-line-numbers-type         t
      projectile-project-search-path    (add-project-directories "~/Documents/Github"
                                                                 "~/Github"
                                                                 "~/Documents/TradingMonkey"
                                                                 "~/Documents/Proyectos/Flutter"
                                                                 "~/Documents/Proyectos/Clojure"
                                                                 "~/Documents/Proyectos/CommonLisp"
                                                                 "~/Documents/Proyectos/Python"
                                                                 "~/Documents/Proyectos/React"
                                                                 "~/Documents/Proyectos/Go"
                                                                 "~/Documents/Proyectos/VHDL"
                                                                 "~/Documents/Proyectos/Machine Learning"
                                                                 "~/Documents/Proyectos/Serverless"
                                                                 "~/Documents"
                                                                 "~/OpenSource")

      evil-vsplit-window-right          t
      evil-split-window-below           t
      doom-font                         (font-spec :family "Fira Code" :size 12)
      display-line-numbers-type         'relative
      ispell-dictionary                 "es"
      ;; doom-themes-treemacs-theme "all-the-icons"
      )

(use-package! po-mode
  :config
  (map! :map po-mode-map
        :localleader
        :desc "Go to next entry" "n" #'po-next-entry
        :desc "Go to prev entry" "N" #'po-previous-entry
        :desc "Next fuzzy entry" "f" #'po-next-fuzzy-entry
        :desc "Prev fuzzy entry" "F" #'po-previous-fuzzy-entry
        :desc "Next untranslated entry" "u" #'po-next-untranslated-entry
        :desc "Prev untranslated entry" "U" #'po-previous-untranslated-entry
        :desc "Edit msgstr in separated buffer" "e" #'po-edit-msgstr))

(use-package! org-tree-slide
  :after org-mode
  :config
  (setq org-tree-slide-slide-in-effect       t
        +org-present-text-scale              2
        org-tree-slide-skip-outline-level    2
        org-tree-slide-fold-subtrees-skipped nil
        org-tree-slide-modeline-display      'outside)
  (map! :map org-tree-slide-mode-map
        :desc "Next slide" "M-n" #'org-tree-slide-move-next-tree
        :desc "Prev slide" "M-p" #'org-tree-slide-move-previous-tree))

(use-package! eca)

(use-package! paredit
  :hook ((clojure-mode . paredit-mode)
         (emacs-lisp-mode . paredit-mode)))

(use-package! lsp-mode
  :commands lsp
  :config
  (setq lsp-headerline-breadcrumb-enable t
        lsp-inlay-hints-mode             t
        lsp-lens-enable                  t)
  (advice-add #'lsp-rename
              :after (lambda (&rest _) (projectile-save-project-buffers))))

(after! lsp-mode
  :config
  (setq lsp-rust-analyzer-display-chaining-hints            t
        lsp-rust-analyzer-display-closure-return-type-hints t
        lsp-rust-analyzer-display-parameter-hints           t))

(use-package! gradle-mode
  :after java-mode)

(use-package! lsp-java
  :after java-mode
  :config
  (add-hook 'java-mode-hook (lambda () (gradle-mode 1)))
  (setq lsp-java-references-code-lens-enabled      t
        lsp-java-implementations-code-lens-enabled t
        lsp-java-lens-mode                         t
        lsp-java-format-settings-url               "~/IDEAJavaFormatter.xml"
        lsp-java-vmargs                            '("-XX:+UseParallelGC"
                                                     "-XX:GCTimeRatio=4"
                                                     "-XX:AdaptiveSizePolicyWeight=90"
                                                     "-Dsun.zip.disableMemoryMapping=true"
                                                     "-Xmx8G"
                                                     "-Xms100m"))
  (setq lombok-library-path (concat doom-data-dir "lombok.jar"))
  (unless (file-exists-p lombok-library-path)
    (url-copy-file "https://projectlombok.org/downloads/lombok.jar" lombok-library-path))
  (push (concat "-javaagent:"
                (expand-file-name lombok-library-path))
        lsp-java-vmargs))

(use-package! lsp-java-boot
  :after lsp-mode
  :preface
  (add-hook 'conf-javaprop-mode-local-vars-hook #'lsp!)
  (add-hook 'yaml-mode-local-vars-hook #'lsp!)
  :config
  (add-hook 'java-mode-local-vars-hook #'lsp-java-boot-lens-mode 'append)
  (add-hook 'yaml-mode-local-vars-hook #'lsp-java-boot-lens-mode 'append))

(use-package! clojure-mode
  :config
  (setq clojure-indent-style 'align-arguments))

(use-package! obsidian
  :config
  (global-obsidian-mode t)
  :custom
  (obsidian-directory "~/Library/Mobile Documents/iCloud~md~obsidian/Documents/ObsNotes")
  (obsidian-daily-notes-directory "Daily"))

(use-package! why-this
  :config
  (setq why-this-annotate-enable-heat-map t
        why-this-idle-delay               0.5
        why-this-minimum-column           45)
  (global-why-this-mode))

(use-package! pulsing-cursor
  :config
  (pulsing-cursor-mode +1))

(use-package! codesnap
  :config
  (map! :localleader
        :prefix "CodeSnap" "S"
        :desc "Screenshot selected code" "S x" #'codesnap-selection
        :desc "Show CodeSnap log" "S l" #'codesnap-show-log))
